File: dev/pilot-camera-validation.mjs
/**
 * Validación P0 cámara piloto Neni — capturas A–G por viewport.
 * Uso: node dev/pilot-camera-validation.mjs [baseUrl]
 * Requiere: npx playwright (se instala bajo demanda).
 */
import { mkdir, writeFile } from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { spawn } from 'node:child_process';

const __dir = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dir, '..');
const OUT = path.join(ROOT, 'docs', 'review', 'pilot-camera-p0');
const BASE = process.argv[2] || 'http://127.0.0.1:8765';

const VIEWPORTS = [
  { id: '740x360', w: 740, h: 360 },
  { id: '844x390', w: 844, h: 390 },
  { id: '932x430', w: 932, h: 430 },
  { id: 'desktop', w: 1366, h: 768 },
];

async function ensurePlaywright() {
  const { chromium } = await import('playwright');
  return chromium;
}

async function maybeStartServer() {
  if (!BASE.includes('127.0.0.1') && !BASE.includes('localhost')) return null;
  const port = Number(new URL(BASE).port || 8765);
  const child = spawn('php', ['-S', `127.0.0.1:${port}`, '-t', ROOT], {
    cwd: ROOT,
    stdio: 'ignore',
    shell: true,
  });
  await new Promise((r) => setTimeout(r, 800));
  return child;
}

async function evalCamera(page) {
  return page.evaluate(() => {
    const svg = document.getElementById('zz-map');
    const st = window.__zz?.getState?.();
    const rect = svg?.getBoundingClientRect?.();
    const vb = svg?.getAttribute('viewBox')?.split(/\s+/).map(Number) || [];
    return {
      viewBox: vb,
      cam: window.__zz?.getState?.()?.mapCamera ? { ...window.__zz.getState().mapCamera } : null,
      svgPx: rect ? { w: rect.width, h: rect.height } : null,
      world: { w: Number(svg?.dataset?.zzWorldW), h: Number(svg?.dataset?.zzWorldH) },
    };
  });
}

async function panToExtreme(page, dir) {
  await page.evaluate(async (direction) => {
    const st = window.__zz?.getState?.();
    const pan = window.__zz?.panBy;
    const clamp = window.__zz?.clampCam;
    if (!st?.mapCamera || !pan) return;
    for (let i = 0; i < 80; i++) {
      const dx = direction === 'left' ? -120 : direction === 'right' ? 120 : 0;
      const dy = direction === 'up' ? -120 : direction === 'down' ? 120 : 0;
      pan(dx, dy);
    }
    clamp?.();
    window.__zz?.paint?.();
  }, dir);
  await page.waitForTimeout(200);
}

async function setZoom(page, mode) {
  await page.evaluate((m) => {
    const st = window.__zz?.getState?.();
    const zoomBy = window.__zz?.zoomBy;
    const clamp = window.__zz?.clampCam;
    if (!st?.mapCamera) return;
    if (m === 'min') {
      const svg = document.getElementById('zz-map');
      const dims = { w: Number(svg?.dataset?.zzWorldW), h: Number(svg?.dataset?.zzWorldH) };
      const r = svg?.getBoundingClientRect?.();
      if (r && dims.w && dims.h) {
        st.mapCamera.zoom = Math.max(r.width / dims.w, r.height / dims.h);
        st.mapCamera.x = dims.w / 2;
        st.mapCamera.y = dims.h / 2;
      }
    } else {
      for (let i = 0; i < 24; i++) zoomBy?.(1.15);
    }
    clamp?.();
    window.__zz?.paint?.();
  }, mode);
  await page.waitForTimeout(200);
}

async function countBlackInMapWrap(page) {
  return page.evaluate(() => {
    const wrap = document.querySelector('.zz-world-map-wrap');
    const svg = document.getElementById('zz-map');
    if (!wrap || !svg) return { error: 'no wrap/svg' };
    const wr = wrap.getBoundingClientRect();
    const sr = svg.getBoundingClientRect();
    const canvas = document.createElement('canvas');
    canvas.width = Math.round(wr.width);
    canvas.height = Math.round(wr.height);
    const ctx = canvas.getContext('2d');
    // Aproximación: muestreo de píxeles del SVG renderizado vía foreignObject no trivial.
    // Marcamos negro si el fondo del wrap (#0c0e10) domina bordes exteriores del SVG.
    const dark = (r, g, b) => r < 30 && g < 30 && b < 35;
    let blackSamples = 0;
    let total = 0;
    const img = svg;
    const vb = svg.getAttribute('viewBox')?.split(/\s+/).map(Number) || [];
    if (vb.length === 4) {
      const [vx, vy, vw, vh] = vb;
      const worldW = Number(svg.dataset.zzWorldW);
      const worldH = Number(svg.dataset.zzWorldH);
      const outside =
        vx < -0.5 || vy < -0.5 || vx + vw > worldW + 0.5 || vy + vh > worldH + 0.5;
      return { outsideViewBox: outside, viewBox: vb, worldW, worldH };
    }
    return { blackSamples, total };
  });
}

async function run() {
  await mkdir(OUT, { recursive: true });
  const server = await maybeStartServer();
  let chromium;
  try {
    chromium = await ensurePlaywright();
  } catch (e) {
    console.error('Instala playwright: npx playwright install chromium');
    throw e;
  }

  const browser = await chromium.launch({ headless: true });
  const report = [];

  for (const vp of VIEWPORTS) {
    const ctx = await browser.newContext({
      viewport: { width: vp.w, height: vp.h },
      deviceScaleFactor: 1,
    });
    const page = await ctx.newPage();
    const url = `${BASE}/play.php?pilot=neni&new=1&clear=1`;
    await page.goto(url, { waitUntil: 'networkidle', timeout: 60000 });
    await page.waitForSelector('#zz-map', { timeout: 30000 });
    await page.waitForTimeout(1500);

    const shots = [
      ['A-d1-inicial', async () => {}],
      ['B-pan-left', async () => panToExtreme(page, 'left')],
      ['C-pan-right', async () => panToExtreme(page, 'right')],
      ['D-pan-up', async () => panToExtreme(page, 'up')],
      ['E-pan-down', async () => panToExtreme(page, 'down')],
      ['F-zoom-min', async () => setZoom(page, 'min')],
      ['G-zoom-max', async () => setZoom(page, 'max')],
    ];

    for (const [label, setup] of shots) {
      if (label !== 'A-d1-inicial') {
        await page.goto(url, { waitUntil: 'networkidle', timeout: 60000 });
        await page.waitForSelector('#zz-map', { timeout: 30000 });
        await page.waitForTimeout(800);
      }
      await setup();
      const meta = await evalCamera(page);
      const black = await countBlackInMapWrap(page);
      const file = `${vp.id}-${label}.png`;
      const wrap = page.locator('.zz-world-map-wrap');
      await wrap.screenshot({ path: path.join(OUT, file) });
      report.push({ viewport: vp.id, shot: label, file, meta, black });
      console.log(`OK ${file}`, black.outsideViewBox === false ? 'in-bounds' : 'OUT-OF-BOUNDS');
    }
    await ctx.close();
  }

  await browser.close();
  if (server) server.kill();
  const reportPath = path.join(OUT, 'report.json');
  await writeFile(reportPath, JSON.stringify(report, null, 2));
  console.log(`\nReport: ${reportPath}`);
  console.log(`Capturas: ${OUT}`);
}

run().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: dev/pilot-camera-validation.py
"""
Validación P0 cámara piloto — recortes del mapa maestro (contrato demo-neni).
Genera capturas A–G por viewport sin Playwright.
"""
import json
import os
from PIL import Image, ImageDraw

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
MAP = os.path.join(
    ROOT,
    "assets",
    "art",
    "terrain",
    "mapa-neni-1819x865-brownmatch-up2x.png",
)
OUT = os.path.join(ROOT, "docs", "review", "pilot-camera-p0")

WORLD_W, WORLD_H = 1819, 865
NUCLEUS_X, NUCLEUS_Y = 1100, 280
ZOOM_MAX = 3.0

VIEWPORTS = [
    ("740x360", 740, 360),
    ("844x390", 844, 390),
    ("932x430", 932, 430),
    ("desktop", 1366, 768),
]


def clamp(v, a, b):
    return max(a, min(b, v))


def zoom_min(vp_w, vp_h):
    return max(vp_w / WORLD_W, vp_h / WORLD_H)


def clamp_cam(cam, vp_w, vp_h):
    zm = zoom_min(vp_w, vp_h)
    cam["zoom"] = clamp(cam["zoom"], zm, ZOOM_MAX)
    half_w = (vp_w / 2) / cam["zoom"]
    half_h = (vp_h / 2) / cam["zoom"]
    vis_w = vp_w / cam["zoom"]
    vis_h = vp_h / cam["zoom"]
    if vis_w >= WORLD_W:
        cam["x"] = WORLD_W / 2
    else:
        cam["x"] = clamp(cam["x"], half_w, WORLD_W - half_w)
    if vis_h >= WORLD_H:
        cam["y"] = WORLD_H / 2
    else:
        cam["y"] = clamp(cam["y"], half_h, WORLD_H - half_h)
    return cam


def view_box(cam, vp_w, vp_h):
    vw = vp_w / cam["zoom"]
    vh = vp_h / cam["zoom"]
    return {
        "x": cam["x"] - vw / 2,
        "y": cam["y"] - vh / 2,
        "w": vw,
        "h": vh,
    }


def render_shot(map_img, vp_w, vp_h, cam, label):
    cam = clamp_cam(dict(cam), vp_w, vp_h)
    vb = view_box(cam, vp_w, vp_h)
    # Mapa up2x → escala lógica 1:1 en coords mundo
    scale = map_img.width / WORLD_W
    x0 = int(round(vb["x"] * scale))
    y0 = int(round(vb["y"] * scale))
    x1 = int(round((vb["x"] + vb["w"]) * scale))
    y1 = int(round((vb["y"] + vb["h"]) * scale))
    crop = map_img.crop((x0, y0, x1, y1)).resize((vp_w, vp_h), Image.LANCZOS)
    draw = ImageDraw.Draw(crop)
    outside = (
        vb["x"] < -0.5
        or vb["y"] < -0.5
        or vb["x"] + vb["w"] > WORLD_W + 0.5
        or vb["y"] + vb["vh"] if False else vb["y"] + vb["h"] > WORLD_H + 0.5
    )
    info = f"{label} z={cam['zoom']:.3f} vb=({vb['x']:.0f},{vb['y']:.0f},{vb['w']:.0f},{vb['h']:.0f})"
    draw.rectangle([0, 0, vp_w - 1, vp_h - 1], outline=(80, 200, 80), width=2)
    draw.rectangle([0, vp_h - 28, vp_w, vp_h], fill=(0, 0, 0))
    draw.text((4, vp_h - 22), info[:90], fill=(180, 240, 180))
    return crop, cam, vb, outside


def main():
    os.makedirs(OUT, exist_ok=True)
    map_img = Image.open(MAP).convert("RGB")
    assert map_img.size[0] / WORLD_W == map_img.size[1] / WORLD_H, map_img.size
    report = []

    for vp_id, vp_w, vp_h in VIEWPORTS:
        zm = zoom_min(vp_w, vp_h)
        d1 = {"x": NUCLEUS_X, "y": NUCLEUS_Y, "zoom": min(ZOOM_MAX, max(zm * 2.5, zm))}
        cam = dict(d1)

        scenarios = [
            ("A-d1-inicial", dict(d1)),
            ("B-pan-left", None),
            ("C-pan-right", None),
            ("D-pan-up", None),
            ("E-pan-down", None),
            ("F-zoom-min", {"x": WORLD_W / 2, "y": WORLD_H / 2, "zoom": zm}),
            ("G-zoom-max", {"x": NUCLEUS_X, "y": NUCLEUS_Y, "zoom": ZOOM_MAX}),
        ]

        for label, preset in scenarios:
            if preset:
                c = dict(preset)
            elif label == "B-pan-left":
                c = {"x": 1e9, "y": NUCLEUS_Y, "zoom": d1["zoom"]}
            elif label == "C-pan-right":
                c = {"x": -1e9, "y": NUCLEUS_Y, "zoom": d1["zoom"]}
            elif label == "D-pan-up":
                c = {"x": NUCLEUS_X, "y": 1e9, "zoom": d1["zoom"]}
            else:
                c = {"x": NUCLEUS_X, "y": -1e9, "zoom": d1["zoom"]}

            img, cam_out, vb, outside = render_shot(map_img, vp_w, vp_h, c, label)
            fname = f"{vp_id}-{label}.png"
            img.save(os.path.join(OUT, fname))
            entry = {
                "viewport": vp_id,
                "shot": label,
                "file": fname,
                "cam": cam_out,
                "viewBox": vb,
                "outsideWorld": outside,
                "zoomMin": zm,
            }
            report.append(entry)
            status = "FAIL outside" if outside else "OK"
            print(f"{status} {fname}")

    with open(os.path.join(OUT, "report.json"), "w", encoding="utf-8") as f:
        json.dump(report, f, indent=2)
    print(f"\nCapturas en {OUT}")


if __name__ == "__main__":
    main()
